### PR TITLE
feat(backend): add GeneralizedCostFunction class

### PR DESCRIPTION
**Pull request by kaligrafy**
Add a new GeneralizedCostFunction class that calculates weighted travel time based on configurable weights for different transit trip components:

- Access/egress travel time with separate origin/destination weights
- In-vehicle travel time with ROW (right-of-way) category weights
- Transfer penalties based on headway frequency thresholds (high/low)
- Headway weights with optional ROW-specific values
- Context-aware frequency thresholds (urban/regional/intercity)

Includes ROWCategory type for type-safe right-of-way classification (A, B, B-, C+, C) and comprehensive value/weight type definitions.

## Summary by CodeRabbit

* **New Features**
  * Added generalized cost calculation for transit routing that computes weighted travel times incorporating penalties for headway, reliability, transfers, and weather conditions, enabling more sophisticated transit analysis and routing optimization.

### CODE REVIEW

File: packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+     * Check if all required values are provided for each leg
+     * @param values GeneralizedCostFunctionValues object
+     * @returns true if all required values are provided for each leg, false otherwise
+     */
+    hasAllRequiredValuesByLeg(values: GeneralizedCostFunctionValues): boolean {
+        if (!values.by_i) {
+            return false;
+        }
+        for (const legIndex in values.by_i) {
+            const leg = values.by_i[legIndex];
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts]: fixed

File: packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+ * Copyright 2026, Polytechnique Montreal and contributors
+ *
+ * This file is licensed under the MIT License.
+ * License text available at https://opensource.org/licenses/MIT
+ */
+
+import TrError from 'chaire-lib-common/lib/utils/TrError';
+import { GeneralizedCostFunctionValues, GeneralizedCostFunctionWeights } from './types';
+
+export class GeneralizedCostFunction {
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts]: set getHeadways to private and removed from tests (tested indirectly). Changed getHeadwayThresholdSeconds to public static so we can use it elsewhere and get the thresholds for documentation in the interface.

File: packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * Get the headway threshold in minutes based on the context.
+     * Services with headway <= threshold are considered "high frequency".
+     * @returns the headway threshold in minutes for the current context
+     */
+    getHeadwayThresholdSeconds(): number {
+        if (this.context === 'urban') {
+            return 10 * 60;
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts]: fixed

File: packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        }
+        return 10 * 60; // default (urban)
+    }
+
+    /**
+     * Calculate the weighted travel time seconds
+     * @param values GeneralizedCostFunctionValues object
+     * @returns the weighted travel time seconds, in seconds
+     */
+    calculateWeightedTravelTimeSeconds(values: GeneralizedCostFunctionValues): number {
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts]: The cost function can be converted to cost, but here we calculate weighted sum of travel time, so I would keep travel time. We may add a real cost function later. but the term GeneralizedCostFunction applies to both weighted time and weighted cost.

File: packages/transition-backend/src/services/generalizedCostFunction/types.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    };
+    transferTravelTimeWeightByMode: {
+        [mode in AccessEgressTransferMode]: number;
+    };
+
+    // In-vehicle travel time weights (w_t_veh):
+    // For ROW, support, vertical alignment and weather protection, the weight for the unknown value must always be set.
+    inVehicleTravelTimeWeightByROW: Record<RightOfWayCategory, number>; // weight for the inVehicleTravelTimeSeconds by ROW category
+    inVehicleTravelTimeWeightBySupport: Record<Support, number>; // weight for the inVehicleTravelTimeSeconds by support (rail, tires, water, suspended, magnetic, air, hover, hydrostatic)
+    inVehicleTravelTimeWeightByVerticalAlignment: Record<VerticalAlignment, number>;
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/types.ts]: added a readme to explain everything

File: packages/transition-backend/src/services/generalizedCostFunction/__tests__/GeneralizedCostFunction.test.ts [hunk 6]
```diff
@@ -0,0 +1,775 @@
...
+            const gcf = new GeneralizedCostFunction(createDefaultWeights(), context);
+            expect(gcf.getHeadwayThresholdSeconds()).toBe(expected);
+        });
+
+        test('should return default 600 seconds for invalid context (defensive code)', () => {
+            // This tests the defensive default return that should never be reached in normal usage
+            // We need to bypass TypeScript's type checking to test this edge case
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            const gcf = new GeneralizedCostFunction(createDefaultWeights(), 'invalid' as any);
+            expect(gcf.getHeadwayThresholdSeconds()).toBe(600);
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/__tests__/GeneralizedCostFunction.test.ts]: obsolete

File: packages/transition-backend/src/services/generalizedCostFunction/__tests__/GeneralizedCostFunction.test.ts [hunk 7]
```diff
@@ -0,0 +1,775 @@
...
+
+        test('should create instance with regional context', () => {
+            const gcf = new GeneralizedCostFunction(createDefaultWeights(), 'regional');
+            expect(gcf.getHeadwayThresholdSeconds()).toBe(15 * 60);
+        });
+
+        test('should create instance with intercity context', () => {
+            const gcf = new GeneralizedCostFunction(createDefaultWeights(), 'intercity');
+            expect(gcf.getHeadwayThresholdSeconds()).toBe(30 * 60);
+        });
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/__tests__/GeneralizedCostFunction.test.ts]: done

File: packages/transition-backend/src/services/generalizedCostFunction/README.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+| $\mu$ | Penalty factor (weight) | - |
+
+## Notes
+
+- **Multiplicative in-vehicle factors**: ROW, support, vertical alignment, and load factor penalties are **multiplied** together (not added)
+- **All-or-nothing headway**: If headway data is missing for any leg, headway costs are excluded for the entire trip
+- **Transfer penalty fallback**: When the transfer index exceeds defined penalties, $c_ {t,\text{max}}$ is used
+- **Weather protection**: Affects waiting time penalties at boarding stops
+- **Load factor**: Added to 1.0 before multiplying with penalty factor to ensure the result is always $\geq 1.0$
+- **Reliability ratio**: On-time performance ratio (0.0 to 1.0), where higher values indicate better reliability
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/README.md]: fixed

File: packages/transition-backend/src/services/generalizedCostFunction/types.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+} from 'transition-common/lib/services/line/types';
+import type { AccessEgressTransferMode } from 'chaire-lib-common/lib/config/routingModes';
+import type { WeatherProtection } from 'transition-common/lib/services/nodes/types';
+
+/** Variables used in the generalized cost function
+ * Symbols and definitions are available at https://www.overleaf.com/read/dtxfhttxgjrx
+ * TODO: implement crowding weights but this will be difficult without having to import demand data.
+ * TODO: implement reliability weights but this will be difficult without having to import reliability data.
+ * TODO: implement a real cost (money) function with values of time and fares
+ */
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/types.ts]: fixed

File: packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            if (headwaysSeconds !== undefined) {
+                const headwaySeconds = headwaysSeconds[i];
+                weightedTravelTimeSeconds +=
+                    headwaySeconds * this.weights.headwayPenaltyWeightByROW[leg.rightOfWayCategory];
+            }
+
+            // Calculate the penalty for reliability ratio:
+            if (leg.reliabilityRatio !== undefined && !isNaN(leg.reliabilityRatio)) {
+                weightedTravelTimeSeconds += this.weights.reliabilityRatioPenaltyWeight * leg.reliabilityRatio;
+            }
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/GeneralizedCostFunction.ts]: fixed, reversed so higher value of reliability gives lower penalties

File: packages/transition-backend/src/services/generalizedCostFunction/types.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    transferPenaltyMax: number;
+
+    /**
+     * Headway penalty weight by ROW category.
+     * Multiplied by headway in seconds.
+     */
+    headwayPenaltyWeightByROW: Record<RightOfWayCategory, number>;
+
+    /**
+     * Penalty weight for unreliability: weight * (1 - reliabilityRatio).
```
Comment (tahini) [inline packages/transition-backend/src/services/generalizedCostFunction/types.ts]: what does it multiply? wait time? travel time? everything?

File: packages/transition-backend/src/services/generalizedCostFunction/types.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+     */
+    waitingTimeWeightByWeatherProtection: Record<WeatherProtection, number>;
+
+    /**
+     * Penalty for each transfer (index starts at 0), in seconds.
+     * E.g., transferPenaltyByIndex[0] = penalty for first transfer.
+     */
+    transferPenaltyByIndex: number[];
+    /**
+     * Maximum penalty for transfer index > provided array length.
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/types.ts]: added

File: packages/transition-backend/src/services/generalizedCostFunction/README.md [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+### Constraints
+
+**All weights ($\mu$) must be $\geq 0$**. This ensures penalties can never be negative:
+- Weights in range $(0, 1)$: reduce perceived time (e.g., productive travel on comfortable train)
+- Weight $= 1$: neutral (no adjustment)
+- Weight $> 1$: increase perceived time (penalty)
+
+**All input time values must be $\geq 0$** (accessTravelTimeSeconds, inVehicleTravelTimeSeconds, etc.).
+
+**Penalties for transfers and hedway must be $\geq 0$** ($c_ {t}^{(j)}$ and $c_ {t,\text{max}}$).
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/generalizedCostFunction/README.md]: fixed
